File: tests/ruby/lifecycle.rb
# lifecycle.rb — covers the hooks and contract corners no other test
# touches: deinit (observable through a log + emit, since the VM is gone
# afterwards), prefab spawning, scene changes, and component remove/has?.

def init
  @marker = Labelle::Entity.create
  @marker.set("Alive")

  # Prefab + scene, including the failure arms.
  ship = Labelle.spawn("ship", x: 5, y: 10)
  raise "spawn failed" if ship.nil?
  ship.set("Tag", kind: "spawned")
  raise "scene change refused" unless Labelle.scene_change("menu")
  raise "unknown scene accepted" if Labelle.scene_change("nope")

  # Remove is idempotent; has? flips accordingly.
  raise "expected Alive" unless @marker.has?("Alive")
  raise "remove failed" unless @marker.remove("Alive")
  raise "still Alive" if @marker.has?("Alive")
  raise "idempotent remove failed" unless @marker.remove("Alive")
end

def deinit
  Labelle.log("ruby: lifecycle deinit ran")
  Labelle.emit("shutdown_done", from: @marker.id)
end

File: tests/rust/game/lifecycle.rs
//! The lifecycle port (lua/lifecycle.lua's rust twin): every remaining
//! contract seam — has/remove, prefab spawn with params, scene change
//! (rejected AND accepted arms), an emit from deinit — plus re-setup
//! semantics: `register` runs afresh, so a second setup builds fresh
//! script state against the same world.

use crate::labelle::{self, EntityId, Script};

#[derive(Default)]
pub struct Lifecycle {
    marker: EntityId,
}

impl Script for Lifecycle {
    fn init(&mut self) {
        self.marker = labelle::create_entity();
        assert!(labelle::set_component(
            self.marker,
            "Alive",
            r#"{"ok":true}"#
        ));
        assert!(labelle::component_has(self.marker, "Alive"));
        assert!(labelle::remove_component(self.marker, "Alive"));
        assert!(!labelle::component_has(self.marker, "Alive"));
        // Removes are idempotent on the component.
        assert!(labelle::remove_component(self.marker, "Alive"));

        let ship =
            labelle::spawn_prefab("ship", Some(r#"{"x":5,"y":10}"#)).expect("prefab spawn failed");
        assert!(labelle::set_component(ship, "Tag", r#"{"kind":"spawned"}"#));

        // The mock's unknown-scene arm refuses; the real one lands.
        assert!(!labelle::change_scene("nope"));
        assert!(labelle::change_scene("menu"));
    }

    fn deinit(&mut self) {
        labelle::log("rust: lifecycle deinit ran");
        labelle::emit("shutdown_done", &format!("{{\"from\":{}}}", self.marker));
    }
}
